### PR TITLE
Isolate flow actions and ensure only them can reference the `buildWorkResult` provider

### PR DESCRIPTION
**Pull request by bamboo**

### CODE REVIEW

**Comment by bamboo** [comment]
@bot-gradle test ACC

File: subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/codecs/ProviderCodecs.kt [hunk 1]
```diff
@@ -144,6 +161,47 @@ class FlowProvidersCodec(
...
+    inline fun <T : MutableIsolateContext, R> T.withFlowActionIsolate(flowActionClass: Class<*>, owner: IsolateOwner.OwnerFlowScope, block: T.() -> R): R {
+        withIsolate(IsolateOwner.OwnerFlowAction(flowActionClass, owner)) {
+            withPropertyTrace(PropertyTrace.BuildLogicClass(flowActionClass.name)) {
+                return block()
+            }
+        }
+    }
+
+    private
+    fun IsolateContext.isolateOwner(): IsolateOwner.OwnerFlowScope {
```
Comment (mlopatkin) [inline subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/codecs/ProviderCodecs.kt:195]: nit: maybe `verifiedIsolateOwner` would stand out a bit more at the call sites. But the class is small enough for it not to matter much.

**Review by mlopatkin — APPROVED** [review]
It looks really nice and clean.

File: subprojects/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/ConfigurationCacheFlowScopeIntegrationTest.groovy [hunk 2]
```diff
@@ -428,6 +478,43 @@ class ConfigurationCacheFlowScopeIntegrationTest extends AbstractConfigurationCa
...
+
+            apply type: FailsPlugin
+        '''
+
+        when:
+        configurationCacheFails 'fails'
+
+        then:
+        failureHasCause "Failed to calculate the value of task ':fails' property 'color'."
+        failureHasCause "Cannot access the value of 'BuildWorkResult' before it becomes available!"
```
[the remark this answers is not in the document]
Comment (bamboo) [reply subprojects/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/ConfigurationCacheFlowScopeIntegrationTest.groovy:514]: Yes.

**Comment by bamboo** [comment]
Thanks for the feedback, @mlopatkin!

**Comment by bamboo** [comment]
@bot-gradle test and merge

**Pull request merged**